### PR TITLE
Adds a type parameter to invokeMethod (and additional utility methods)

### LINKED ISSUE #17333: Add a type argument `<T>` to `invokeMethod`
i.e. instead of:
https://github.com/flutter/flutter/blob/0a4c1166833083815d34a4a963a093ea1f63bf11/packages/flutter/lib/src/services/platform_channel.dart#L271-L280

Something like:

```dart
Future<T> invokeMethod<T>(String method, [dynamic arguments]) async {
  ...
  return result as T;
}
```

There are a lot of runtime errors being reported by errors:
https://github.com/dart-lang/pub-dartlang-dart/issues/1263#issuecomment-386845927

The advantage here is that `T` can be filled in automatically in most cases:

```dart
// Automatically infers invokeMethod<bool>!
Future<bool> isEnabled() => invokeMethod('isEnabled');
```

... and I imagine this would be a non-breaking change, as a plus! This won't work for more complex types, like `List<String>`, but maybe that's either OK _or_ some variants of `invokeMethod` could be added to help, for example:
* `Future<List<T>> invokeMethodList<T>(...)`
* `Future<Map<K, V>> invokeMethodMap<K, V>(...)`

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/17333

This adds a type parameter T to the `invokeMethod` method, which allows it to automatically return a properly typed Future instead of requiring a manual cast. Also adds `invokeListMethod` and `invokeMapMethod` utilities for Lists and Maps, respectively. Due to the reified nature of Dart generics, it is not possible to cast a Future, List, or Map without creating a new wrapper. 

Before:

```
// API returns a bool
final result = channel.invokeMethod('example');
acceptsFutureBool(result); // Might crash, Future<dynamic> cannot be assigned to Future<bool>.
```

After:

```
final result = channel.invokeMethod<bool>('example');
// or
final bool result = channel.invokeMethod('example'); // T is inferred
acceptsFutureBool(result); // yay!
```

Additionally, I've added two utility methods to help with returning Maps and List (which have similar issues):

Before:

```
// Api returns a Map<String, String>, but actually it is Map<dynamic, dynamic>
final Map<String, String> result = await channel.invokeMethod('example'); // Map<dynamic, dynamic> cannot be assigned to Map<String, String>
```
```
final Map<String, String> result = await channel.invokeMethod('example').cast<String, String>();
```

After:
```
final result = await channel.invokeMapMethod<String, String>('example');
// or
final Map<String, String> result = await channel.invokeMapMethod('example'); // T is inferred.
```

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I guess @optionalTypeArguments isnt that optional

File: packages/flutter/lib/src/services/haptic_feedback.dart [hunk 1]
```diff
@@ -21,7 +21,7 @@ class HapticFeedback {
   /// On Android, this uses the platform haptic feedback API to simulate a
   /// response to a long press (`HapticFeedbackConstants.LONG_PRESS`).
   static Future<void> vibrate() async {
-    await SystemChannels.platform.invokeMethod('HapticFeedback.vibrate');
+    return SystemChannels.platform.invokeMethod('HapticFeedback.vibrate');
```
Comment (Hixie) [inline packages/flutter/lib/src/services/haptic_feedback.dart]: return seems conceptually wrong here. We're in an `async` function, so we should return a real value, not a Future. (I know that technically an async body has a return value of `FutureOr<T>`, not `T`, but it seems conceptually wrong to do that IMHO.)

File: packages/flutter/lib/src/services/platform_channel.dart [hunk 2]
```diff
@@ -128,6 +128,12 @@ class MethodChannel {
   /// result. The values supported by the default codec and their platform-specific
   /// counterparts are documented with [StandardMessageCodec].
   ///
+  /// The generic argument `T` of the method can be inferred by the surrounding
+  /// context, or provided explicitly. If it does not match the returned type of
+  /// the channel, a [TypeError] will be thrown at runtime. `T` cannot be class
+  /// with generics, such as `Map<String, String>`, instead it must always
+  /// return an untyped map or list.
```
Comment (Hixie) [inline packages/flutter/lib/src/services/platform_channel.dart]: untyped meaning what? `Map<dynamic, dynamic>` isn't "untyped", it's just typed with dynamic...

File: packages/flutter/lib/src/services/haptic_feedback.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/services/haptic_feedback.dart]: The nice thing about this one is that type inference gives you `void` for free, but I don't have a strong opinion myself. happy to change it back.

**Comment by Hixie** [comment]
LGTM modulo comments (in particular, I don't think we should replace the awaits for void into returns)

File: packages/flutter/lib/src/services/haptic_feedback.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/services/haptic_feedback.dart]: Done

File: packages/flutter/lib/src/services/platform_channel.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/services/platform_channel.dart]: Re-phrased to clarify that only dynamic is supported

**Pull request merged**